### PR TITLE
DeviceManager refactor 4: Define testDevice and DeviceCombo to operate devices

### PR DESCRIPTION
**Pull request by zhou9584**
## Description

Define testDevice and DeviceCombo to operate devices
## Testing

Trigger Espresso,ADB, E2E, T2C Test

### CODE REVIEW

File: common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java [hunk 1]
```diff
@@ -131,6 +131,7 @@ public boolean uninstallApp(DeviceInfo deviceInfo, String packageName, Logger lo
 
     @Override
     public void resetPackage(DeviceInfo deviceInfo, String packageName, Logger logger) {
+        uninstallApp(deviceInfo, packageName, logger);
         classLogger.info("Nothing Implemented for iOS in " + currentMethodName());
```
Comment (dinghong2233) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java:135]: remove this line?

Comment (zhou9584) [reply common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java:135]: Rewrite this method. Will not verify the implementation of TestDeviceManager is iOS anymore.
image

File: common/src/main/java/com/microsoft/hydralab/common/management/device/TestDevice.java [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+
+public class TestDevice implements TestDeviceOperation {
+    private final DeviceInfo deviceInfo;
+    private final TestDeviceTag tag;
```
Comment (zhou9584) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/TestDevice.java:24]: update Enum type to string

File: common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceOperation.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+import com.microsoft.hydralab.common.entity.common.TestRun;
+import com.microsoft.hydralab.common.entity.common.TestTask;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+
+interface TestDeviceOperation {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceOperation.java:16]:
```suggestion
interface TestDeviceCapability {
```

File: common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceTag.java [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// Licensed under the MIT License.
+
+package com.microsoft.hydralab.common.management.device;
+
+/**
+ * @author zhoule
+ * @date 03/13/2023
+ */
+
+public enum TestDeviceTag {
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceTag.java:11]: make this string constants?

File: common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java [hunk 5]
```diff
@@ -131,7 +131,8 @@ public boolean uninstallApp(DeviceInfo deviceInfo, String packageName, Logger lo
 
     @Override
     public void resetPackage(DeviceInfo deviceInfo, String packageName, Logger logger) {
-        classLogger.info("Nothing Implemented for iOS in " + currentMethodName());
+        // todo Nothing Implemented for iOS in resetPackage replace with uninstallApp
+        uninstallApp(deviceInfo, packageName, logger);
```
Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java:135]:
```suggestion

```

Comment (hydraxman) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/impl/IOSTestDeviceManager.java:135]:
```suggestion
        uninstallApp(deviceInfo, packageName, logger);
        classLogger.info("!!!  As there is no implementation for package reset in iOS, what we did here is uninstall the app, please be AWARE  !!!");
```

File: agent/doc/UML/agent_device_management_design.puml [hunk 6]
```diff
@@ -1,30 +1,79 @@
...
+}
+
+
+class DeviceInfo
+interface ITestDeviceCapability
+class TestDeviceCapability implements ITestDeviceCapability
+class TestDeviceComboCapability implements ITestDeviceCapability
+class TestDeviceCapabilityAdapter implements ITestDeviceCapability
+
+class TestDevice{
```
Comment (zhou9584) [inline agent/doc/UML/agent_device_management_design.puml:36]: ITestTun 持有 testDevice

Comment (zhou9584) [inline agent/doc/UML/agent_device_management_design.puml:36]: TestRunDevice 
TestRun .device

File: agent/doc/UML/agent_device_management_design.puml [hunk 7]
```diff
@@ -1,30 +1,79 @@
...
+    onDeviceActive(DeviceInfo)
+}
+class TestTaskEngineService{
+    TestDevice chooseDevice(TestTask)
+    Map<String, TestDevice> taskDeviceMap
+}
+
+
+class DeviceInfo
+interface ITestDeviceCapability
```
Comment (zhou9584) [inline agent/doc/UML/agent_device_management_design.puml:31]: 定义方法

File: agent/doc/UML/agent_device_management_design.puml [hunk 8]
```diff
@@ -1,30 +1,79 @@
...
+ScreenRecorder screenRecorder;
+LogCollector logCollector;
+Map<Thread, String> currentCommand
+Map<Thread, String> currentProcess
+Map<Thread, TestTask> currentTask
+}
+
+class TestDeviceCombo extends TestDevice{
+List<TestDevice> pairedTestDevices
+TestDevice getDeviceByTag(String tag)
```
Comment (zhou9584) [inline agent/doc/UML/agent_device_management_design.puml:48]: getAllTestDevices()

File: common/src/main/java/com/microsoft/hydralab/common/management/device/TestDevice.java [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+import com.microsoft.hydralab.common.util.ThreadPoolUtil;
+import com.microsoft.hydralab.common.util.ThreadUtils;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+
+public class TestDevice implements TestDeviceCapability {
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/TestDevice.java:22]: If the TestSession no longer contains the operation, SingleDeviceSession is preferred for me.
Session implies that the lifecycle is the same as TestTask.

SingleDeviceSession, MultiDeviceSession.

File: common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceCapability.java [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+import com.microsoft.hydralab.common.entity.common.TestRun;
+import com.microsoft.hydralab.common.entity.common.TestTask;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.slf4j.Logger;
+
+import java.io.File;
+import java.io.IOException;
+
+interface TestDeviceCapability {
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/common/management/device/TestDeviceCapability.java:16]: TestDeviceCapability -> TestDeviceOrchestrator

File: agent/doc/UML/agent_device_management_design.puml (hunk 8 above)
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:48]:
```suggestion
TestDevice getDeviceByTag(String tag)
List<TestDevice> getDevices()
```

File: agent/doc/UML/agent_device_management_design.puml [hunk 11]
```diff
@@ -1,30 +1,79 @@
...
+class TestTaskEngineService{
+    TestDevice chooseDevice(TestTask)
+    Map<String, TestDevice> taskDeviceMap
+}
+
+
+class DeviceInfo
+interface ITestDeviceCapability
+class TestDeviceCapability implements ITestDeviceCapability
+class TestDeviceComboCapability implements ITestDeviceCapability
```
Comment (zhou9584) [inline agent/doc/UML/agent_device_management_design.puml:33]: 扁平化TestDeviceCapability

Comment (AC-Phoenix) [reply agent/doc/UML/agent_device_management_design.puml:33]: 也许可以从删除TestDeviceCapabilityAdapter入手

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java [hunk 12]
```diff
@@ -6,22 +6,46 @@
...
 import com.microsoft.hydralab.common.util.FileUtil;
 import com.microsoft.hydralab.common.util.ThreadPoolUtil;
-import com.microsoft.hydralab.performance.inspectors.*;
-import com.microsoft.hydralab.performance.parsers.*;
+import com.microsoft.hydralab.performance.inspectors.AndroidBatteryInfoInspector;
+import com.microsoft.hydralab.performance.inspectors.AndroidMemoryInfoInspector;
+import com.microsoft.hydralab.performance.inspectors.IOSEnergyGaugeInspector;
+import com.microsoft.hydralab.performance.inspectors.IOSMemoryPerfInspector;
+import com.microsoft.hydralab.performance.inspectors.WindowsBatteryInspector;
+import com.microsoft.hydralab.performance.inspectors.WindowsMemoryInspector;
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java:17]:
```suggestion
import com.microsoft.hydralab.performance.inspectors.*;
```

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java [hunk 13]
```diff
@@ -6,22 +6,46 @@
...
+import com.microsoft.hydralab.performance.parsers.AndroidBatteryInfoResultParser;
+import com.microsoft.hydralab.performance.parsers.AndroidMemoryInfoResultParser;
+import com.microsoft.hydralab.performance.parsers.IOSEnergyGaugeResultParser;
+import com.microsoft.hydralab.performance.parsers.IOSMemoryPerfResultParser;
+import com.microsoft.hydralab.performance.parsers.WindowsBatteryResultParser;
+import com.microsoft.hydralab.performance.parsers.WindowsMemoryResultParser;
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java:23]:
```suggestion
import com.microsoft.hydralab.performance.parsers.*;
```

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java [hunk 14]
```diff
@@ -6,22 +6,46 @@
...
 import org.slf4j.Logger;
 import org.springframework.util.Assert;
 
 import java.io.File;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java:33]:
```suggestion
import java.util.*;
```

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java [hunk 15]
```diff
@@ -6,22 +6,46 @@
...
 import java.util.concurrent.ScheduledFuture;
 
-import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.*;
-import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.*;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_ANDROID_BATTERY_INFO;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_ANDROID_MEMORY_INFO;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_IOS_ENERGY;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_IOS_MEMORY;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_WIN_BATTERY;
+import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.INSPECTOR_WIN_MEMORY;
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java:42]:
```suggestion
import static com.microsoft.hydralab.performance.PerformanceInspector.PerformanceInspectorType.*;
```

File: common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java [hunk 16]
```diff
@@ -6,22 +6,46 @@
...
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_ANDROID_BATTERY_INFO;
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_ANDROID_MEMORY_INFO;
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_IOS_ENERGY;
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_IOS_MEMORY;
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_WIN_BATTERY;
+import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.PARSER_WIN_MEMORY;
```
Comment (AC-Phoenix) [inline common/src/main/java/com/microsoft/hydralab/performance/PerformanceTestManagementService.java:48]:
```suggestion
import static com.microsoft.hydralab.performance.PerformanceResultParser.PerformanceResultParserType.*;
```

**Comment by AC-Phoenix** [comment]
These three names often make me feel confused, at least from the name I can't tell what their duties are.
- TestDeviceManager (Long-term life cycle)
- TestRunDeviceOrchestrator (Long-term life cycle)
- TestRunDevice (The life cycle is the same as the test task)

I think we should be careful to name a class as a `Manger`, because TestDeviceManager seems to manage only one device. It seems that it does certain things `Driver `does. Can we merge it into the corresponding `Driver`?

The prefixes of `TestRunDeviceOrchestrator` and `TestRunDevice` are the same, but the life cycle seems to be different. When I see both classes at the same time, I assume that their lifecycles are the same.
I still think `SingleTestRunDeviceSession `is a better name for `TestRunDevice `, because it is very much like a session that is created when a test task is triggered. In the case of multiple devices, another subclass can be named as `MultiTestRunDeviceSession `

File: agent/doc/UML/agent_device_management_design.puml [hunk 17]
```diff
@@ -1,32 +1,103 @@
...
+}
+Class TestRun{
+    TestRunDevice device
+}
+class TestRunner{
+    runTest(TestTask,TestRunDevice)
+}
+
+interface IDeviceDriver
+class DeviceDriverManager implements IDeviceDriver{
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:78]: Please add
```
DeviceDriverManager *--> AndroidDeviceDriver
DeviceDriverManager *--> IOSDeviceDriver   
DeviceDriverManager *--> WindowsDeviceDriver 
```

File: agent/doc/UML/agent_device_management_design.puml [hunk 18]
```diff
@@ -1,32 +1,103 @@
...
+
+class TestRunDevice{
+    String Tag
+    DeviceInfo deviceInfo;
+    WebDriver appiumDriver;
+    ScreenRecorder screenRecorder;
+    LogCollector logCollector;
+    Map<Thread, String> currentCommand
+    Map<Thread, String> currentProcess
+    Map<Thread, TestTask> currentTask
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:62]: I think these 3 should be put into device info layer.

File: agent/doc/UML/agent_device_management_design.puml [hunk 19]
```diff
@@ -1,32 +1,103 @@
 @startuml
 'https://plantuml.com/class-diagram
+enum DeviceType{
+    ANDROID
+    IOS
+    WINDOWS
+}
+
+class DeviceInfo{
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:9]:
```suggestion
class ConnectedDevice{
```

File: agent/doc/UML/agent_device_management_design.puml [hunk 20]
```diff
@@ -1,32 +1,103 @@
...
+}
+class TestTaskEngineService{
+    TestRunDevice chooseDevice(TestTask)
+}
+
+
+class DeviceInfo
+
+class TestRunDeviceOrchestrator{
+    void setup(TestRunDevice, Logger);
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:32]: Should TestRun be a parameter of the `setup` and `unset`?

File: agent/doc/UML/agent_device_management_design.puml [hunk 21]
```diff
@@ -1,32 +1,103 @@
...
-WindowsTestDeviceManager --|> TestDeviceManager
+TestRunDeviceOrchestrator --> DeviceDriverManager
 
-AgentManagementService *--> AndroidTestDeviceManager
-AgentManagementService *--> IOSTestDeviceManager
-AgentManagementService *--> WindowsTestDeviceManager
+TestTaskEngineService --> TestRunner
 
-TestDeviceManager --> TestDriver
+IDeviceDriver -> DeviceInfo
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:94]: Could you also list the method of `IDeviceDriver`? Like what we did with `TestRunDeviceOrchestrator`.

File: agent/doc/UML/agent_device_management_design.puml [hunk 22]
```diff
@@ -1,32 +1,103 @@
...
+    List<TestRunDevice> getDevices()
+}
+Class TestRun{
+    TestRunDevice device
+}
+class TestRunner{
+    runTest(TestTask,TestRunDevice)
+}
+
+interface IDeviceDriver
```
Comment (hydraxman) [inline agent/doc/UML/agent_device_management_design.puml:77]:
```suggestion
interface DeviceDriver
```

File: agent/doc/UML/agent_device_management_design.puml (hunk 20 above)
Comment (zhou9584) [reply agent/doc/UML/agent_device_management_design.puml:32]: 感觉如果把TestRun加进来的话，就跟testRunner 里的setUp方法有职责混淆，我的想法是这里只做device相关的配置
这几个方法可以包一个
image

**Review by hydraxman — APPROVED** [review]
UML design approved, not sure about detail quality.

**Pull request merged**